myriad-scheduler/src/main/java/com/ebay/myriad/scheduler/MyriadOperations.java

```diff
@@ -37,103 +37,124 @@
+          Set<NodeTask> activeTasksForProfile = Sets.newHashSet(this.schedulerState.getActiveTasksForProfile(profile));
+          List<String> nodesToScaleDown = nodeScaleDownPolicy.getNodesToScaleDown();
+          filterUnregisteredNMs(activeTasksForProfile, nodesToScaleDown);
 
-            for (Protos.TaskID taskId : stagingTasks) {
-                this.schedulerState.makeTaskKillable(taskId);
-                numScaledDown++;
-                if (numScaledDown == numInstancesToScaleDown) {
-                    break;
+          for (int i = 0; i < numInstancesToScaleDown - (numPendingTasksScaledDown + numStagingTasksScaledDown); i++) {
+            for (NodeTask nodeTask : activeTasksForProfile) {
+              if (nodesToScaleDown.size() > i &&
+                  nodesToScaleDown.get(i).equals(nodeTask.getHostname()) &&
+                  meetsConstraint(nodeTask, constraint)) {
+                this.schedulerState.makeTaskKillable(nodeTask.getTaskStatus().getTaskId());
+                numActiveTasksScaledDown++;
+                if (LOGGER.isDebugEnabled()) {
+                  LOGGER.debug("Marked NodeTask {} on host {} for kill.",
+                      nodeTask.getTaskStatus().getTaskId(), nodeTask.getHostname());
                 }
+              }
             }
+          }
         }
-        int numStagingTasksScaledDown = numScaledDown - numActiveTasksScaledDown;
 
-        // Flex down Pending tasks, if any
-        if (numScaledDown < numInstancesToScaleDown) {
-            Set<Protos.TaskID> pendingTasks = Sets.newHashSet(this.schedulerState.getPendingTaskIds());
+        if (numActiveTasksScaledDown + numStagingTasksScaledDown + numPendingTasksScaledDown == 0) {
+          LOGGER.info("No Node Managers with profile '{}' and constraint {} found for scaling down.",
+              profile.getName(), constraint.toString());
+        } else {
+          LOGGER.info("Flexed down {} active, {} staging  and {} pending Node Managers with '{}' profile.",
+              numActiveTasksScaledDown, numStagingTasksScaledDown, numPendingTasksScaledDown, profile.getName());
+        }
+    }
 
-            for (Protos.TaskID taskId : pendingTasks) {
-                this.schedulerState.makeTaskKillable(taskId);
-                numScaledDown++;
-                if (numScaledDown == numInstancesToScaleDown) {
-                    break;
-                }
+  private boolean meetsConstraint(NodeTask nodeTask, Constraint constraint) {
+    if (constraint != null) {
+      if (constraint.equals(nodeTask.getConstraint())) {
+        return true;
+      }
+      switch (constraint.getType()) {
+        case LIKE:
+          LikeConstraint likeConstraint = (LikeConstraint) constraint;
+          if (likeConstraint.isConstraintOnHostName()) {
+            return likeConstraint.matchesHostName(nodeTask.getHostname());
+          } else {
+            return likeConstraint.matchesSlaveAttributes(nodeTask.getSlaveAttributes());
+          }
+      }
```

Comment:
Oh yes. I think it makes sense to have a ```default: return false;``` to imply that the currently supported constraint is just LIKE. Will add it.